### PR TITLE
Give 'typeof'-like errors in attribute type arguments

### LINKED ISSUE #54778: Generic attributes should disallow type arguments which are disallowed by 'typeof'
**Version Used**: 3661e8354e3887e9826300171080c28e0d241a8e

**Steps to Reproduce**:

```cs
using System;

class Attr1 : Attribute
{
    public Attr1(Type t) { }
}

class Attr2<T> : Attribute { }

[Attr1(typeof(dynamic))]
[Attr2<dynamic>]
class C { }
```

**Expected Behavior**: Error on `typeof(dynamic)` as well as on `Attr2<dynamic>`.

**Actual Behavior**: Error only on `typeof(dynamic)`.

We want to do this because when we emit attribute type arguments there's nowhere to put things like `DynamicAttribute`, for example. So the fact that the type argument is `dynamic` cannot round trip.

LDM has tentatively expressed support for the change. We should still probably create a speclet in csharplang for this feature so that all these behaviors are documented.

### LINKED ISSUE #54804: Warn when "attribute-dependent" types are used in 'typeof'
Currently the 'typeof' operator gives an error when certain types which can't round-trip are used as arguments. For example, `dynamic` or `string?`.

[SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4GIYFcA2OLA4CmABETAcQLABQtAAgAwn0CMALANy20D6JAXhIwiAdwDaAXRIBvWiQUkALgE8ADkQD2AMwAUm4ACsiAYyUBKOCQD01kpoDW8xao07dCFRTBQTlm3ZEAE5BmkEkAMIAyqwAnABsAEwgJAAqABakrlra9hpBEEphJCYQMDCaSiTApFgAzkQI9jDKmSSe3r7K6kTOCtnubIwA/P62ZCHF0QAc8QDMsSkZWT06ecGFxaXlldW1DU2aLRDCuPiEpEFE2sHkJisafd1uegAyUHVKADwdED4mAD4xnYKhNQkEngM3h9vkNhkCrONQcFwZDVnpdLAqhArFjquZzAFhJowWFaABfThAA==)
```cs
#nullable enable

using System.Collections.Generic;

_ = new[] {
    typeof(object), // ok
    typeof(dynamic), // error CS1962: The typeof operator cannot be used on the dynamic type
    typeof(string?), // error CS8639: The typeof operator cannot be used on a nullable reference type
    typeof(List<dynamic>), // no error
    typeof(List<string?>), // no error
    typeof((int a, int b)) // no error
};
```

It feels like the lines marked `no error` should be errors because the types in use can't round trip. Since it's not practical to add new errors to existing code, these scenarios should probably be handled in a warning wave.

Generic attribute type arguments have a similar issue where they can't round-trip types which require extra attributes. It feels like the same restriction should be applied there, but perhaps as an error since it's a new feature. #54778

### PR DESCRIPTION
**Pull request by RikkiGibson**
Resolves #54804
Resolves #54778
Related to #36285

### CODE REVIEW

**Comment by jcouv** [comment]
>                 diagnostics.Add(ErrorCode.ERR_FunctionPointerTypesInAttributeNotSupported, node.Location);

Function pointers can't be used a type arguments, correct? Consider adding a test

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:1298 in 30b5f98. [](commit_id = 30b5f982b5cd4ca580e5085aa172cde48aa60ba5, deletion_comment = False)

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -1317,6 +1303,27 @@ private BoundExpression BindTypeOf(TypeOfExpressionSyntax node, BindingDiagnosti
             return new BoundTypeOfOperator(node, boundType, null, this.GetWellKnownType(WellKnownType.System_Type, diagnostics, node), hasError);
         }
 
+        /// <summary>Called when an "attribute-dependent" type such as 'dynamic', 'string?', etc. is not permitted.</summary>
```
[the remark this answers is not in the document]
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:1306]: To-do in #54804

[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:1306]: I think we should just go ahead with diagnostic in generic attribute scenario. For `typeof`, those will only be warning-wave warnings.

[the remark this answers is not in the document]
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:1306]: I think it's the same work. It's just we would check `isTypeof` and give a typeof-specific warning diagnostic versus an error diagnostic for generic attribute arguments. I'm fine with just getting it done though.

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests.cs [hunk 2]
```diff
@@ -9938,62 +9938,59 @@ static void Main()
 
         [Fact]
         [WorkItem(54778, "https://github.com/dotnet/roslyn/issues/54778")]
-        public void GenericAttribute_Dynamic_01()
+        public void GenericAttribute_AttributeDependentTypes()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests.cs:9941]: Consider testing a nested situation (type is bad in multiple ways) to show that we only report once.
For example `List<string?>?` or `(int a, (int b, int c))`.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2)

**Comment by RikkiGibson** [comment]
@jaredpar let me know if you have any concerns about the changes since your last review.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 6)

**Comment by roji** [comment]
Just to let people know that this broke EF Core (and apparently also Orchard, /cc @sebastienros), see https://github.com/dotnet/efcore/pull/25664. In our case this was the use of `typeof(object?[])`, which is indeed meaningless and can be rewritten to `typeof(object[])`, but could be said to convey more information when reading the code... Regardless, just to make sure the potential breakage is known.

**Pull request merged**